sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DirectStreamObserver.java

```diff
@@ -34,21 +38,66 @@
  */
 @ThreadSafe
 public final class DirectStreamObserver<T> implements StreamObserver<T> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DirectStreamObserver.class);
+  private static final int DEFAULT_MAX_MESSAGES_BEFORE_CHECK = 100;
+
   private final Phaser phaser;
   private final CallStreamObserver<T> outboundObserver;
+  private final int maxMessagesBeforeCheck;
+
+  private int numberOfMessagesBeforeReadyCheck;
 
   public DirectStreamObserver(
       Phaser phaser,
       CallStreamObserver<T> outboundObserver) {
+    this(phaser, outboundObserver, DEFAULT_MAX_MESSAGES_BEFORE_CHECK);
+  }
+
+  DirectStreamObserver(
+      Phaser phaser,
+      CallStreamObserver<T> outboundObserver,
+      int maxMessagesBeforeCheck) {
     this.phaser = phaser;
     this.outboundObserver = outboundObserver;
+    this.maxMessagesBeforeCheck = maxMessagesBeforeCheck;
   }
 
   @Override
   public void onNext(T value) {
-    int phase = phaser.getPhase();
-    if (!outboundObserver.isReady()) {
-      phaser.awaitAdvance(phase);
+    numberOfMessagesBeforeReadyCheck += 1;
+    if (numberOfMessagesBeforeReadyCheck >= maxMessagesBeforeCheck) {
+      numberOfMessagesBeforeReadyCheck = 0;
+      int waitTime = 1;
+      int totalTimeWaited = 0;
```

Comment:
The 'wait-to-advance' seems like it should be in its own method.

I'm not super happy with this logic being in the StreamObserver - perhaps it can be in a utility? It looks really similar to a backoff.

Reply:
The numberOfMessagesBeforeReadyCheck is stateful with context to the class so moving it to another utility is just going to add boiler plate and a level of redirection that won't be helpful.